### PR TITLE
CA-2052 Add safety check to reboot after 30 minutes of no GSM connection

### PR DESCRIPTION
**Pull request by Tooseriuz**
### Changes

- Add service for checking connectivity every 1.30 minutes
- if connectivity timeout for 15 minutes, the device will reboot

### CODE REVIEW

**Comment by antonyharfield** [comment]
Sorry @Tooseriuz, we will hold this for now as @topherwhite has a better solution that reuses the failure checks in the ApiCheckInService.

**Comment by Tooseriuz** [comment]
@antonyharfield from the diff that you provided in Jira, we do not need this PR anymore if that is working.
But the problem is that we don't know on the Bourhan's state that it can detect the sim or not

File: core/src/main/java/org/rfcx/guardian/utility/device/DeviceMobilePhone.java [hunk 1]
```diff
@@ -76,7 +76,16 @@ private String getDeviceIMSI() {
...
+	}
+
+	private Boolean getSimState() {
+		TelephonyManager tm = (TelephonyManager) this.context.getSystemService(Context.TELEPHONY_SERVICE);
+		return !(tm.getSimState() == TelephonyManager.SIM_STATE_ABSENT);
+	}
+
+	public Boolean hasSim() {
+		return getSimState();
+	}
```
Comment (antonyharfield) [inline core/src/main/java/org/rfcx/guardian/utility/device/DeviceMobilePhone.java]: I don't understand why these are 2 separate methods. Isn't it better as 1?

File: core/src/main/java/org/rfcx/guardian/utility/device/DeviceCPU.java [hunk 2]
```diff
@@ -69,29 +69,31 @@ private void incrementClockSpeedAverage() {
...
+				reader.close();
+				toks = load.split(" ");
+				long idle2 = Long.parseLong(toks[5]);
+				long cpu2 = Long.parseLong(toks[2]) + Long.parseLong(toks[3]) + Long.parseLong(toks[4])
+						+ Long.parseLong(toks[6]) + Long.parseLong(toks[7]) + Long.parseLong(toks[8]);
+				return (float) (cpu2 - cpu1) / ((cpu2 + idle2) - (cpu1 + idle1));
+			} catch (IOException e) {
+				RfcxLog.logExc(logTag, e);
+			}
+		}
```
Comment (antonyharfield) [inline core/src/main/java/org/rfcx/guardian/utility/device/DeviceCPU.java]: These lines have not changed... try not to commit them.

Use the technique I suggested before:

```
if (!DeviceRoot.isRooted()) {
   return 0;
}
```

File: admin/src/main/java/org/rfcx/guardian/admin/activity/MainActivity.kt [hunk 3]
```diff
@@ -89,7 +89,9 @@ class MainActivity : AppCompatActivity() {
 
     public override fun onPause() {
         super.onPause()
-        monitorTimer.cancel()
+        if(::monitorTimer.isInitialized){
```
Comment (antonyharfield) [inline admin/src/main/java/org/rfcx/guardian/admin/activity/MainActivity.kt]: Style.

File: admin/src/main/java/org/rfcx/guardian/admin/RfcxGuardian.java [hunk 4]
```diff
@@ -86,6 +87,7 @@ public void onCreate() {
 		ShellCommands.triggerNeedForRootAccess(this);
 		DeviceI2cUtils.resetI2cPermissions(this);
 		DateTimeUtils.resetDateTimeReadWritePermissions(this);
+		DeviceBluetooth.setOn();
```
Comment (antonyharfield) [inline admin/src/main/java/org/rfcx/guardian/admin/RfcxGuardian.java]: As commented before, should be inside an if to check the pref.

**Review by antonyharfield — CHANGES_REQUESTED** [review]

**Review by antonyharfield — APPROVED** [review]
Looks good now.

**Review by topherwhite — APPROVED** [review]
looks good to me and thanks for using the thresholds!

**Pull request merged**